server/auth/lib.py

```diff
@@ -65,211 +62,118 @@ def is_superadmin():
-            if key not in kwargs:
-                raise Exception(
-                    f"expected '{key}' in kwargs but got: {kwargs}"
-                )  # pragma: no cover
-
-        election = get_or_404(Election, kwargs.pop("election_id"))
-        jurisdiction = get_or_404(Jurisdiction, kwargs.pop("jurisdiction_id"))
-
-        require_jurisdiction_admin_for_jurisdiction(jurisdiction.id, election.id)
-
-        kwargs["election"] = election
-        kwargs["jurisdiction"] = jurisdiction
-
-        return route(*args, **kwargs)
-
-    return wrapper
-
-
-def with_audit_board_access(route: Callable):
-    """
-    Flask route decorator that restricts access to a route to Audit Board
-    members that are part of the audit board for the election, jurisdiction,
-    round, and audit board ids in the route's path. It also loads the
-    election, jurisdiction, round, and audit board objects.
-
-    To use this, you must have:
-    - a path component named `election_id`
-    - a route parameter named `election`
-    - a path component named `jurisdiction_id`
-    - a route parameter named `jurisdiction`
-    - a path component named `round_id`
-    - a route parameter named `round`
-    - a path component named `audit_board_id`
-    - a route parameter named `audit_board`
-    """
-
-    @functools.wraps(route)
-    def wrapper(*args, **kwargs):
-        for key in ["election_id", "jurisdiction_id", "round_id", "audit_board_id"]:
-            if key not in kwargs:
-                raise Exception(
-                    f"expected '{key}' in kwargs but got: {kwargs}"
-                )  # pragma: no cover
-
-        election = get_or_404(Election, kwargs.pop("election_id"))
-        jurisdiction = get_or_404(Jurisdiction, kwargs.pop("jurisdiction_id"))
-        round = get_or_404(Round, kwargs.pop("round_id"))
-        audit_board = get_or_404(AuditBoard, kwargs.pop("audit_board_id"))
-
-        require_audit_board_logged_in(audit_board, election.id, jurisdiction, round.id)
-
-        kwargs["election"] = election
-        kwargs["jurisdiction"] = jurisdiction
-        kwargs["round"] = round
-        kwargs["audit_board"] = audit_board
-
-        return route(*args, **kwargs)
-
-    return wrapper
-
-
-def with_superadmin_access(route: Callable):
+    def restrict_access_decorator(route: Callable):
+        @functools.wraps(route)
+        def wrapper(*args, **kwargs):
+            # Substitute route params for their corresponding resources
+            if "election_id" in kwargs:
+                election = get_or_404(Election, kwargs.pop("election_id"))
+                kwargs["election"] = election
+
+            jurisdiction = None
+            if "jurisdiction_id" in kwargs:
+                jurisdiction = find_or_404(
+                    Jurisdiction.query.filter_by(
+                        id=kwargs.pop("jurisdiction_id"), election_id=election.id
+                    )
+                )
```

Comment:
this will fail if the path has `jurisdiction_id` but not `election_id`, right? Maybe fail explicitly in that case?

Reply:
ea499a5